staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go

```diff
@@ -71,3 +71,104 @@ func Convert_v1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s
+		return err
+	}
+
+	if len(out.Versions) == 0 && len(in.Version) > 0 {
+		// no versions were specified, and a version name was specified
+		out.Versions = []CustomResourceDefinitionVersion{{Name: in.Version, Served: true, Storage: true}}
+	}
+
+	// If spec.{subresources,validation,additionalPrinterColumns} exists, move to versions
+	if in.Subresources != nil {
+		subresources := &CustomResourceSubresources{}
+		if err := Convert_apiextensions_CustomResourceSubresources_To_v1_CustomResourceSubresources(in.Subresources, subresources, s); err != nil {
+			return err
+		}
+		for i := range out.Versions {
+			out.Versions[i].Subresources = subresources
+		}
+	}
+	if in.Validation != nil {
+		schema := &CustomResourceValidation{}
+		if err := Convert_apiextensions_CustomResourceValidation_To_v1_CustomResourceValidation(in.Validation, schema, s); err != nil {
+			return err
+		}
+		for i := range out.Versions {
+			out.Versions[i].Schema = schema
+		}
+	}
+	if in.AdditionalPrinterColumns != nil {
+		additionalPrinterColumns := make([]CustomResourceColumnDefinition, len(in.AdditionalPrinterColumns))
+		for i := range in.AdditionalPrinterColumns {
+			if err := Convert_apiextensions_CustomResourceColumnDefinition_To_v1_CustomResourceColumnDefinition(&in.AdditionalPrinterColumns[i], &additionalPrinterColumns[i], s); err != nil {
+				return err
+			}
+		}
+		for i := range out.Versions {
+			out.Versions[i].AdditionalPrinterColumns = additionalPrinterColumns
+		}
+	}
+	return nil
+}
+
+func Convert_v1_CustomResourceDefinitionSpec_To_apiextensions_CustomResourceDefinitionSpec(in *CustomResourceDefinitionSpec, out *apiextensions.CustomResourceDefinitionSpec, s conversion.Scope) error {
+	if err := autoConvert_v1_CustomResourceDefinitionSpec_To_apiextensions_CustomResourceDefinitionSpec(in, out, s); err != nil {
+		return nil
+	}
+
+	if len(out.Versions) == 0 {
+		return nil
+	}
+
+	// Copy versions[0] to version
+	out.Version = out.Versions[0].Name
+
+	// If versions[*].{subresources,schema,additionalPrinterColumns} are identical, move to spec
+	subresources := out.Versions[0].Subresources
+	subresourcesIdentical := true
+	validation := out.Versions[0].Schema
+	validationIdentical := true
+	additionalPrinterColumns := out.Versions[0].AdditionalPrinterColumns
+	additionalPrinterColumnsIdentical := true
+
+	// Detect if per-version fields are identical
+	for _, v := range out.Versions {
+		if subresourcesIdentical && !apiequality.Semantic.DeepEqual(v.Subresources, subresources) {
+			subresourcesIdentical = false
+		}
+		if validationIdentical && !apiequality.Semantic.DeepEqual(v.Schema, validation) {
+			validationIdentical = false
+		}
+		if additionalPrinterColumnsIdentical && !apiequality.Semantic.DeepEqual(v.AdditionalPrinterColumns, additionalPrinterColumns) {
+			additionalPrinterColumnsIdentical = false
+		}
+	}
+
+	// If they are, set the top-level fields and clear the per-version fields
```

Comment:
do we already have internal validation that ensures that this condition holds (one or the other)?

Reply:
yes. https://github.com/kubernetes/kubernetes/blob/release-1.15/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L183-L192